### PR TITLE
[IMP] test_lint,*: prevent usage of %r in translation strings

### PR DESCRIPTION
**Pull request by MissingNoShiny**
While using %r in translation strings to automatically quote strings without having to deal with escaping/choosing the right string delimiters seems like a good idea, it has a few issues:

- It's intended for debugging purposes.
- It doesn't work for non-string data types.
- Different languages use various quotation marks (e.g., “guillemets” in French, corner brackets in Chinese).
- Even in English, the common typographical recommendations suggest using curly quotes (“”) over straight quotes ("").
- It is not always clear whether it will produce single or double quotes.

For these reasons, this PR removes all instances of %r in translation strings, and replaces them with %s, surrounding them with curly quotes.

Enterprise: https://github.com/odoo/enterprise/pull/63270

Task-3942743

### CODE REVIEW

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -719,13 +719,13 @@ def _check_depends(self):
...
                         raise UserError(_(
-                            'Unknown field "%(field)s" in dependency "%(dependency)s"',
+                            'Unknown field “%(field)s” in dependency “%(dependency)s”',
                             field=name,
                             dependency=seq.strip(),
                         ))
                     if index < last and not field.relational:
                         raise UserError(_(
-                            'Non-relational field "%(field)s" in dependency "%(dependency)s"',
+                            'Non-relational field “%(field)s” in dependency “%(dependency)s”',
```
Comment (oomsveta) [inline odoo/addons/base/models/ir_model.py:728]: Unrelated diff?

Comment (MissingNoShiny) [reply odoo/addons/base/models/ir_model.py:728]: Indeed, I just thought it would be weird if just that particular error message does not have curly quotes, and it was changed last week in the placeholders PRs, so 🤷‍♂️

File: odoo/addons/test_lint/tests/_odoo_checker_gettext.py [hunk 2]
```diff
@@ -43,9 +43,14 @@ class OdooBaseChecker(BaseChecker):
             'gettext-placeholders',
             'Use keyword arguments when you have multiple placeholders',
         ),
+        'E8506': (
+            'Usage of %r in _, _lt function',
+            'gettext-repr',
+            'Don\'t use %r to automatically insert quotes in translation strings',
```
Comment (oomsveta) [inline odoo/addons/test_lint/tests/_odoo_checker_gettext.py:49]: I'd add a little explanation, "Quotes can be different depending on the language, they must be part of the translated string."

File: addons/account_edi_proxy_client/models/account_edi_proxy_user.py [hunk 3]
```diff
@@ -134,7 +134,7 @@ def _make_request(self, url, params=False):
         if 'error' in response:
             message = _('The url that this service requested returned an error. The url it tried to contact was %(url)s. %(error_message)s', url=url, error_message=response['error']['message'])
             if response['error']['code'] == 404:
-                message = _('The url that this service tried to contact does not exist. The url was %r', url)
+                message = _('The url that this service tried to contact does not exist. The url was “%s”', url)
```
Comment (ticodoo) [inline addons/account_edi_proxy_client/models/account_edi_proxy_user.py:137]: I think this sounds better:
```suggestion
                message = _('The url that this service tried to contact does not exist. The url in question is “%s”', url)
```

That being said, the error messages above this all have the same `was` wording that sounds a little strange to me so :shrug: probably doesn't matter

**Review by ticodoo — COMMENTED** [review]
My best guesses. I still think that the curl quotes should not be used since it will result in inconsistent curl vs non-curly quotes within our strings.

Also, I didn't understand a lot of the base strings, so some of them could probably be better but it's hard to reword them + ensure that they are still accurate so I didn't touch them unless it was an obvious grammar mistake

**Comment by HydrionBurst** [comment]
I think the changes are OK, but I hold a different idea for the commit message.
I think the %r is not used for auto quote, `%r` is designed to return a string in a way that can be used to recreate the original object. In short `%r` is for developers and `%s` is for users.
Sementically `_("%(field)r", field)` means you need to debug the field, and there might be something wrong. In that case you shouldn't replace it with `%s` even if they return the same result.
For python `str`, `%r` sometimes is helpful than `%s`. e.g. `'%s' % "aa\tbb" == "aa\tbb"`, `'%r' % "aa\tbb" == ""'aa\\tbb'""` It would be easier for developers to debug.

I think in most cases `%s` is good and sementically correct for the message. But maybe you can recheck if the message is used to debug.

File: odoo/addons/test_lint/tests/_odoo_checker_gettext.py [hunk 4]
```diff
@@ -54,6 +59,8 @@ def visit_call(self, node):
                     self.add_message('gettext-variable', node=node)
                 elif len(PLACEHOLDER_REGEXP.findall(str(first_arg.value))) >= 2:
                     self.add_message('gettext-placeholders', node=node)
+                elif re.search(r"%(\(\w+\))?r", first_arg.value):
```
Comment (Julien00859) [inline odoo/addons/test_lint/tests/_odoo_checker_gettext.py:62]: compile the regexp, move the compiled object on top and change it to use non-capturing group
```py
r"%(?:\(\w+\))?r"
```

**Comment by MissingNoShiny** [comment]
@HydrionBurst A lot of the `%r` were introduced in PR https://github.com/odoo/odoo/pull/139314 with the specific purpose of auto-quoting, that's why I phrased the commit message that way. Good point though, that's another reason to not use them, I will add it to the list

**Comment by Julien00859** [comment]
It is going to be a burden to use those curly quotes in new i18n text. I suggest we add a comment inside `odoo/tools/translate.py` at the top of the file with the two curly quotes for easy copy/paste. It'll then be easy to spread the word "please use the curly quotes, you can find them in translate.py" in future reviews.

I suggest:
```diff
diff --git a/odoo/tools/translate.py b/odoo/tools/translate.py
index e41e88d18a23..e005f4cfcce1 100644
--- a/odoo/tools/translate.py
+++ b/odoo/tools/translate.py
@@ -1,5 +1,11 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+# Please use “curly quotes” instead of "double quotes" in English text.
+#
+# On Linux the keyboard shortcut for unicode character is shift+alt+u+<code>
+# “ (opening curly quote) has code 201c
+# ” (closing curly quote) has code 201d
+
 import codecs
 import fnmatch
 import functools
```

**Comment by MissingNoShiny** [comment]
On Linux, it's even simpler than that, you can use AltGr + V/B :slightly_smiling_face:

**Comment by oomsveta** [comment]
robodoo r+

**Comment by Julien00859** [comment]
@robodoo r-

checking something real fast, sorry

**Comment by Julien00859** [comment]
@robodoo r+ sorry for the fuss, everything is fine, blame @d-fence
